pennylane/gradients/hadamard_gradient.py

```diff
@@ -0,0 +1,318 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions for computing the parameter-shift gradient
+of a qubit-based quantum tape.
+"""
+import pennylane as qml
+import pennylane.numpy as np
+from pennylane.measurements import MutualInfoMP, StateMP, VarianceMP, VnEntropyMP
+from pennylane.transforms.metric_tensor import _get_aux_wire
+from pennylane.transforms.tape_expand import expand_invalid_trainable_hadamard_gradient
+from .finite_difference import _all_zero_grad_new, _no_trainable_grad_new
+
+from .gradient_transform import (
+    choose_grad_methods,
+    grad_method_validation,
+    gradient_analysis,
+    gradient_transform,
+)
+
+
+def _hadamard_grad(
+    tape,
+    argnum=None,
+    shots=None,
+    aux_wire=None,
+    device_wires=None,
+):
+    r"""Transform a QNode to compute the hadamard test gradient of all gate
+    parameters with respect to its inputs. This function is adapted to the new return system.
```

Comment:
Please could you add some references?

Reply:
Yes references and well written documentation will be added :+1: